examples/cpp/helloworld/greeter_async_bidi_client.cc

```diff
@@ -0,0 +1,188 @@
+    stream_->Read(&response_, reinterpret_cast<void*>(Type::READ));
+  }
+
+  // Runs a gRPC completion-queue processing thread. Checks for 'Next' tag
+  // and processes them until there are no more (or when the completion queue
+  // is shutdown).
+  void GrpcThread() {
+    while (true) {
+      void* got_tag;
+      bool ok = false;
+      // Block until the next result is available in the completion queue "cq".
+      // The return value of Next should always be checked. This return value
+      // tells us whether there is any kind of event or the cq_ is shutting
+      // down.
+      if (!cq_.Next(&got_tag, &ok)) {
+        std::cerr << "Client stream closed. Quitting" << std::endl;
+        break;
+      }
+
+      // It's important to process all tags even if the ok is false. One might
+      // want to deallocate memory that has be reinterpret_cast'ed to void*
+      // when the tag got initialized. For our example, we cast an int to a
+      // void*, so we don't have extra memory management to take care of.
+      if (ok) {
+        std::cout << std::endl
+                  << "**** Processing completion queue tag " << got_tag
+                  << std::endl;
+        switch (static_cast<Type>(reinterpret_cast<long>(got_tag))) {
+          case Type::READ:
+            std::cout << "Read a new message." << std::endl;
+            break;
+          case Type::WRITE:
+            std::cout << "Sending message (async)." << std::endl;
+            AsyncHelloRequestNextMessage();
+            break;
+          case Type::CONNECT:
+            std::cout << "Server connected." << std::endl;
+            break;
+          default:
+            std::cerr << "Unexpected tag " << got_tag << std::endl;
+            GPR_ASSERT(false);
+        }
+      }
+    }
+  }
+
+  // Context for the client. It could be used to convey extra information to
+  // the server and/or tweak certain RPC behaviors.
+  ClientContext context_;
+
+  // The producer-consumer queue we use to communicate asynchronously with the
+  // gRPC runtime.
+  CompletionQueue cq_;
+
+  // Out of the passed in Channel comes the stub, stored here, our view of the
+  // server's exposed services.
+  std::unique_ptr<MultiGreeter::Stub> stub_;
+
+  // The bidirectional, asynchronous stream for sending/receiving messages.
+  std::unique_ptr<ClientAsyncReaderWriter<HelloRequest, HelloReply>> stream_;
+
+  // Allocated protobuf that holds the response. In real clients and servers,
+  // the memory management would a bit more complex as the thread that fills
+  // in the response should take care of concurrency as well as memory
+  // management.
+  HelloReply response_;
+
+  // Thread that notifies the gRPC completion queue tags.
+  std::unique_ptr<std::thread> grpc_thread_;
+};
+
+int main(int argc, char** argv) {
+  AsyncBidiGreeterClient greeter(grpc::CreateChannel(
+      "localhost:50051", grpc::InsecureChannelCredentials()));
+
+  std::string user;
```

Comment:
Which block? If it's the while block, then it won't work as I would repeatedly create multiple clients..